Add context-specific status

kylechui opened the pull request:
Just added a small if statement that updates the `editing_text` variable depending on whether the buffer is modifiable or not.

P.S. This is my first PR, so please let me know if there's any "PR etiquette" or similar I should be following. Thanks!

andweeb commented:
Hi @kylechui, thanks for the PR! Nah this is great :) I'll review this now!

andweeb commented on lua/presence/init.lua:
```diff
@@ -466,6 +466,12 @@ function Presence:update_for_buffer(buffer, should_debounce)
+        editing_text = "Editing %s"
+    else
+        editing_text = "Reading %s"
+    end
```
Currently, the `editing_text` option can be either a static format string or a function with the signature `function(filename, buffer): string`.

Since we're moving away from a static `"Editing %s"` string to a dynamic one, it'd be nice to have this logic in some `get_editing_text` function to be set on the option instead.

What do you think about moving these lines to a new method and updating the existing [`set_option`](https://github.com/kylechui/presence.nvim/blob/main/lua/presence/init.lua#L91) call to use that method as the default value argument? Pseudocode would probably be something like:
```lua
function Presence.get_editing_text(filename)
    if modifiable and not readonly then
        return string.format("Editing %s", filename)
    else
        return string.format("Reading %s", filename)
    end
end

function Presence:setup(options)
    -- ... existing setup code...
    self:set_option("editing_text", self.get_editing_text)
    -- ... more setup code...
end
```

kylechui replied:
Sorry, I saw this---I think a new method sounds like the best way of going about things. I'll try implementing this right now

andweeb replied:
Hmm actually that's a good idea, but considering that people may already be using the `editing_text` option I wouldn't want to release a breaking change without first having some intermediary deprecation warning.

That may be a bit out of scope for this feature, so we can keep it as `editing_text` for now and I'll implement deprecation warnings after this feature's merged in!

kylechui commented:
Sorry about working on the main branch, I didn't realise it would automatically add my commits to the PR. As for the latest commit, I added a new function `get_status_text` that does the dynamic status text. So far I've added support for non-modifiables (`Reading %s`) and a few file tree explorers (`Browsing %s`).

kylechui commented on lua/presence/init.lua:
```diff
@@ -466,6 +466,12 @@ function Presence:update_for_buffer(buffer, should_debounce)
     }
 
     local editing_text = self.options.editing_text
+    if vim.bo.modifiable then
```
Is this true? It would be trivial to add, but from my own testing it already works perfectly fine with help files.

andweeb replied:
Right, for help files `vim.bo.modifiable` is set to `false` so your change works perfectly for any unmodifiable file whether it's a help file or some file opened using the `-M` option. For read-only or immutable files however, `vim.bo.modifiable` is actually set to `true` since you can still edit the files, but run into warnings when either editing and saving the file:
```sh
$ touch read-only.txt      # Create a new file
$ chmod 0444 read-only.txt # Make it a read-only file
$ nvim read-only.txt       # Discord will show "Editing read-only.txt"
```

I've been internally debating and I can go either way on this since _technically_ you can be editing the file, so I'll resolve this but feel free to let me know what you think!

kylechui replied:
I think that if the majority of time you're reading instead of editing these immutable files, then it should default to "reading". Users can set `vim.bo.modifiable` to true if they want the "editing" text message

andweeb commented on lua/presence/init.lua:
```diff
@@ -465,13 +477,8 @@ function Presence:update_for_buffer(buffer, should_debounce)
-
     local activity = {
-        state = editing_text,
+        state = self.get_status_text(filename),
```
We actually don't need any changes in this function now that we're setting the `get_status_text` method on the `editing_text` option in the `setup` function.

Adding the method call here would actually remove the ability for users to configure their own text. :) You can revert it like this to make the `editing_text` now work as expected:

```suggestion
    local editing_text = self.options.editing_text
    editing_text = type(editing_text) == "function"
         and editing_text(filename, buffer)
         or string.format(editing_text, filename)

    local activity = {
        state = editing_text,
```

kylechui replied:
Oh I see how that works now, just reverted it

andweeb replied:
Oof I think I botched the suggestion 🤦 Can you remove that extranous `local activity = {` text at line 486?

kylechui replied:
Yeah I already got it in the latest commit :)

andweeb replied:
O that was quick, thanks! 🚀

andweeb reviewed (approved):
LGTM 🚀 Thanks @kylechui!
